File: socorre_ai_partner/lib/core/utils/partner_document_rules.dart
import 'partner_type_utils.dart';

class PartnerDocumentDefinition {
  final String type;
  final String title;
  final String description;
  final bool required;

  const PartnerDocumentDefinition({
    required this.type,
    required this.title,
    required this.description,
    required this.required,
  });
}

class PartnerDocumentRules {
  static const String rgCpf = 'rg_cpf';
  static const String cnh = 'cnh';
  static const String crlv = 'crlv';
  static const String cnpj = 'cnpj';
  static const String addressProof = 'address_proof';
  static const String businessLicense = 'business_license';
  static const String certification = 'certification';
  static const String workshopPhotos = 'workshop_photos';
  static const String vehiclePhotos = 'vehicle_photos';
  static const String insurancePolicy = 'insurance_policy';
  static const String environmentalLicense = 'environmental_license';
  static const String fireSafety = 'fire_safety';
  static const String stationPhotos = 'station_photos';
  static const String storePhotos = 'store_photos';
  static const String inventoryPhotos = 'inventory_photos';
  static const String towPhotos = 'tow_photos';

  static List<PartnerDocumentDefinition> getAll(String partnerType) {
    switch (PartnerTypeUtils.normalize(partnerType)) {
      case PartnerTypeUtils.mechanic:
        return const [
          PartnerDocumentDefinition(
            type: rgCpf,
            title: 'RG, CPF ou CNH',
            description: 'Documento de identificação com foto do responsável',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: addressProof,
            title: 'Comprovante de Residência',
            description: 'Conta de luz, água ou telefone',
            required: true,
          ),
        ];
      case PartnerTypeUtils.motoboy:
        return const [
          PartnerDocumentDefinition(
            type: rgCpf,
            title: 'RG ou CPF',
            description: 'Documento de identidade com foto',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: cnh,
            title: 'CNH',
            description: 'Carteira Nacional de Habilitação',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: crlv,
            title: 'CRLV',
            description: 'Documento do veículo',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: addressProof,
            title: 'Comprovante de Residência',
            description: 'Conta de luz, água ou telefone',
            required: true,
          ),
        ];
      case PartnerTypeUtils.tow:
        return const [
          PartnerDocumentDefinition(
            type: rgCpf,
            title: 'RG ou CPF',
            description: 'Documento de identidade com foto',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: cnh,
            title: 'CNH',
            description: 'Carteira Nacional de Habilitação adequada',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: crlv,
            title: 'CRLV',
            description: 'Documento do guincho',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: addressProof,
            title: 'Comprovante de Residência',
            description: 'Conta de luz, água ou telefone',
            required: true,
          ),
        ];
      case PartnerTypeUtils.gasStation:
        return const [
          PartnerDocumentDefinition(
            type: rgCpf,
            title: 'RG ou CPF do Responsável',
            description: 'Documento do responsável pelo posto',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: cnpj,
            title: 'Cartão CNPJ',
            description: 'Comprovante de inscrição do CNPJ',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: businessLicense,
            title: 'Alvará de Funcionamento',
            description: 'Licença municipal de funcionamento',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: addressProof,
            title: 'Comprovante de Endereço',
            description: 'Comprovante do endereço comercial',
            required: true,
          ),
        ];
      case PartnerTypeUtils.autoParts:
        return const [
          PartnerDocumentDefinition(
            type: rgCpf,
            title: 'RG ou CPF do Responsável',
            description: 'Documento do responsável pela loja',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: cnpj,
            title: 'Cartão CNPJ',
            description: 'Comprovante de inscrição do CNPJ',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: businessLicense,
            title: 'Alvará de Funcionamento',
            description: 'Licença municipal de funcionamento',
            required: true,
          ),
          PartnerDocumentDefinition(
            type: addressProof,
            title: 'Comprovante de Endereço',
            description: 'Comprovante do endereço comercial',
            required: true,
          ),
        ];
      default:
        return const [];
    }
  }

  static List<String> getRequiredTypes(String partnerType) {
    return getAll(partnerType)
        .where((document) => document.required)
        .map((document) => document.type)
        .toList();
  }

  static bool requiresDocuments(String partnerType) {
    return getRequiredTypes(partnerType).isNotEmpty;
  }

  static String normalizeDocumentType(String documentType) {
    switch (documentType) {
      case 'cpf':
      case 'identidade':
        return rgCpf;
      case 'vehicle_document':
        return crlv;
      case 'residence_proof':
        return addressProof;
      default:
        return documentType;
    }
  }

  static bool matchesType(String left, String right) {
    return normalizeDocumentType(left) == normalizeDocumentType(right);
  }
}

File: socorre_ai_backend/src/config/partnerDocumentRules.js
const CANONICAL_PARTNER_TYPES = {
  mechanic: 'mechanic',
  motoboy: 'motoboy',
  gas_station: 'gas_station',
  auto_parts: 'auto_parts',
  tow: 'tow',
};

const PARTNER_TYPE_ALIASES = {
  gasstation: CANONICAL_PARTNER_TYPES.gas_station,
  store: CANONICAL_PARTNER_TYPES.gas_station,
  autoparts: CANONICAL_PARTNER_TYPES.auto_parts,
  towtruck: CANONICAL_PARTNER_TYPES.tow,
  tow_truck: CANONICAL_PARTNER_TYPES.tow,
  guincho: CANONICAL_PARTNER_TYPES.tow,
  delivery: CANONICAL_PARTNER_TYPES.motoboy,
};

const DOCUMENT_TYPE_ALIASES = {
  cpf: 'rg_cpf',
  identidade: 'rg_cpf',
  vehicle_document: 'crlv',
  residence_proof: 'address_proof',
};

const DOCUMENT_RULES = {
  [CANONICAL_PARTNER_TYPES.mechanic]: [
    {
      type: 'rg_cpf',
      title: 'RG, CPF ou CNH',
      required: true,
    },
    {
      type: 'address_proof',
      title: 'Comprovante de Residência',
      required: true,
    },
  ],
  [CANONICAL_PARTNER_TYPES.motoboy]: [
    {
      type: 'rg_cpf',
      title: 'RG ou CPF',
      required: true,
    },
    {
      type: 'cnh',
      title: 'CNH',
      required: true,
    },
    {
      type: 'crlv',
      title: 'CRLV',
      required: true,
    },
    {
      type: 'address_proof',
      title: 'Comprovante de Residência',
      required: true,
    },
  ],
  [CANONICAL_PARTNER_TYPES.tow]: [
    {
      type: 'rg_cpf',
      title: 'RG ou CPF',
      required: true,
    },
    {
      type: 'cnh',
      title: 'CNH',
      required: true,
    },
    {
      type: 'crlv',
      title: 'CRLV',
      required: true,
    },
    {
      type: 'address_proof',
      title: 'Comprovante de Residência',
      required: true,
    },
  ],
  [CANONICAL_PARTNER_TYPES.gas_station]: [
    {
      type: 'rg_cpf',
      title: 'RG ou CPF do Responsável',
      required: true,
    },
    {
      type: 'cnpj',
      title: 'Cartão CNPJ',
      required: true,
    },
    {
      type: 'business_license',
      title: 'Alvará de Funcionamento',
      required: true,
    },
    {
      type: 'address_proof',
      title: 'Comprovante de Endereço',
      required: true,
    },
  ],
  [CANONICAL_PARTNER_TYPES.auto_parts]: [
    {
      type: 'rg_cpf',
      title: 'RG ou CPF do Responsável',
      required: true,
    },
    {
      type: 'cnpj',
      title: 'Cartão CNPJ',
      required: true,
    },
    {
      type: 'business_license',
      title: 'Alvará de Funcionamento',
      required: true,
    },
    {
      type: 'address_proof',
      title: 'Comprovante de Endereço',
      required: true,
    },
  ],
};

function normalizePartnerType(partnerType) {
  if (!partnerType) {
    return '';
  }

  return PARTNER_TYPE_ALIASES[partnerType] || partnerType;
}

function normalizeDocumentType(documentType) {
  if (!documentType) {
    return '';
  }

  return DOCUMENT_TYPE_ALIASES[documentType] || documentType;
}

function getAllDocuments(partnerType) {
  return DOCUMENT_RULES[normalizePartnerType(partnerType)] || [];
}

function getRequiredDocuments(partnerType) {
  return getAllDocuments(partnerType)
    .filter((document) => document.required)
    .map((document) => document.type);
}

function requiresDocuments(partnerType) {
  return getRequiredDocuments(partnerType).length > 0;
}

function determineDocumentTypeFromFilename(filename) {
  const lowerFilename = (filename || '').toLowerCase();

  if (lowerFilename.includes('cpf') || lowerFilename.includes('rg') || lowerFilename.includes('identidade')) {
    return 'rg_cpf';
  }
  if (lowerFilename.includes('cnpj')) {
    return 'cnpj';
  }
  if (lowerFilename.includes('cnh') || lowerFilename.includes('habilitacao')) {
    return 'cnh';
  }
  if (lowerFilename.includes('crlv') || lowerFilename.includes('veiculo') || lowerFilename.includes('vehicle_document')) {
    return 'crlv';
  }
  if (
    lowerFilename.includes('residencia') ||
    lowerFilename.includes('endereco') ||
    lowerFilename.includes('comprovante') ||
    lowerFilename.includes('address_proof') ||
    lowerFilename.includes('residence_proof')
  ) {
    return 'address_proof';
  }
  if (lowerFilename.includes('certificado') || lowerFilename.includes('certification')) {
    return 'certification';
  }
  if (lowerFilename.includes('licenca') || lowerFilename.includes('alvara') || lowerFilename.includes('business_license')) {
    return 'business_license';
  }
  if (lowerFilename.includes('ambiental')) {
    return 'environmental_license';
  }
  if (lowerFilename.includes('bombeiro') || lowerFilename.includes('incendio') || lowerFilename.includes('fire')) {
    return 'fire_safety';
  }
  if (lowerFilename.includes('seguro') || lowerFilename.includes('insurance')) {
    return 'insurance_policy';
  }

  return 'other';
}

module.exports = {
  normalizePartnerType,
  normalizeDocumentType,
  getAllDocuments,
  getRequiredDocuments,
  requiresDocuments,
  determineDocumentTypeFromFilename,
  CANONICAL_PARTNER_TYPES,
};
